### PR TITLE
Implement GCS upload in datasets

### PR DESCRIPTION
**Pull request by mociarain**
This PR contains the functionality to upload files and streams to GCS

### CODE REVIEW

**Comment by acroz** [comment]
@mociarain This can now be rebased off master.

**Comment by acroz** [comment]
@mociarain I've done the rebase myself for the sake of saving time.

File: tests/datasets/test_transfer.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+
+from faculty.clients.object import CloudStorageProvider, CompletedUploadPart
+from faculty.datasets import transfer
+
+
+PROJECT_ID = uuid4()
+TEST_PATH = "/path/to/file"
+TEST_URL = "https://example.com/presigned/url"
+OTHER_URL = "https://example.com/other-presigned/url"
+TEST_AWS_UPLOAD_ID = 123
```
Comment (acroz) [inline tests/datasets/test_transfer.py]: S3 upload IDs are strings.

Can we also:
```suggestion
TEST_S3_UPLOAD_ID = 123
```

File: tests/datasets/test_transfer.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+@pytest.fixture
+def mock_client_upload_s3(mocker, requests_mock):
+    presigned_response_mock = mocker.Mock()
+    presigned_response_mock.provider = CloudStorageProvider.S3
+    presigned_response_mock.upload_id = TEST_AWS_UPLOAD_ID
+
+    object_client = mocker.Mock()
+    object_client.presign_upload.return_value = presigned_response_mock
+
+    yield object_client
```
Comment (acroz) [inline tests/datasets/test_transfer.py]: If we don't want this fixture to do anything after the test finishes, it does not need to be a generator and we can just:
```suggestion
    return object_client
```
However, we may want to keep it as a generator and add the following:
```python
object_client.presign_upload.assert_called_once_with(PROJECT_ID, TEST_PATH)
```

File: tests/datasets/test_transfer.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+@pytest.fixture
+def mock_client_upload_gcs(mocker, requests_mock):
+    presigned_response_mock = mocker.Mock()
+    presigned_response_mock.provider = CloudStorageProvider.GCS
+    presigned_response_mock.url = TEST_URL
+
+    object_client = mocker.Mock()
+    object_client.presign_upload.return_value = presigned_response_mock
+
+    yield object_client
```
Comment (acroz) [inline tests/datasets/test_transfer.py]: As above.

File: tests/datasets/test_transfer.py [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+    transfer.download_file(
+        mock_client_download, PROJECT_ID, TEST_PATH, destination
+    )
+    assert destination.read(mode="rb") == TEST_CONTENT
+
+
+def test_chunking_and_labelling_data_of_exact_sizes(mocker):
+    mocker.patch("faculty.datasets.transfer.UPLOAD_CHUNK_SIZE", 4)
+    content = [b"1111", b"2222", b"last"]
+    a = transfer._rechunk_and_label_as_last(content)
```
Comment (acroz) [inline tests/datasets/test_transfer.py]: "a" isn't a great variable name. Can we change it to something like "chunks"?

File: tests/datasets/test_transfer.py [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+    assert list(a) == [(b"1111", False), (b"2222", False), (b"last", True)]
+
+
+def test_chunking_and_labelling_of_smaller_sizes(mocker):
+    mocker.patch("faculty.datasets.transfer.UPLOAD_CHUNK_SIZE", 12)
+    content = [b"1111", b"2222", b"last"]
+    a = transfer._rechunk_and_label_as_last(content)
```
Comment (acroz) [inline tests/datasets/test_transfer.py]: As above:
"a" isn't a great variable name. Can we change it to something like "chunks"?

File: tests/datasets/test_transfer.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+    assert list(a) == [(b"11112222last", True)]
+
+
+def test_chunking_and_labelling_of_greater_sizes(mocker):
+    mocker.patch("faculty.datasets.transfer.UPLOAD_CHUNK_SIZE", 4)
+    content = [b"11112222last"]
+    a = transfer._rechunk_and_label_as_last(content)
```
Comment (acroz) [inline tests/datasets/test_transfer.py]: As above:
"a" isn't a great variable name. Can we change it to something like "chunks"?

File: tests/datasets/test_transfer.py (hunk 2 above)
Comment (mociarain) [reply tests/datasets/test_transfer.py]: I think I just forgot the assertion. I'll add that now

**Pull request merged**